### PR TITLE
Credits loading sanatization

### PR DESCRIPTION
**Pull request by Hop311**
- checks to avoid crashes on missing or incorrectly formatted credits files.
- alphabetic sorting of people within roles.

### CODE REVIEW

**Review by BrickPi — APPROVED** [review]
Good work.

File: game/src/Credits/Credits.gd [hunk 1]
```diff
@@ -32,72 +32,85 @@ var label_settings_personel : LabelSettings
...
+	label.name = 'Label' + text
+	label.text = text
+	label.horizontal_alignment = HORIZONTAL_ALIGNMENT_CENTER
+	label.label_settings = settings
+	node.add_child(label)
+
 # REQUIREMENTS:
 # * UI-34, UI-35
-func make_project_credits(project):
+func _add_project_credits(project) -> void:
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:95]: Why are you adding a underscore prefix to the identifier? Afaik this should only apply to the methods that are related to signals and other such godot engine related methods. The [gd script style guide](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html) suggests that methods should just be snake case.

It also has no whitespace between the identifier and colon for static typing

```suggestion
func add_project_credits(project) -> void:
```

File: game/src/Credits/Credits.gd [hunk 2]
```diff
@@ -32,72 +32,85 @@ var label_settings_personel : LabelSettings
...
+
 # REQUIREMENTS:
 # * 1.5 Credits Menu
 # * SS-17
 
 # REQUIREMENTS
 # * FS-4
-func load_credit_file(path):
-	var roles = {}
+func _load_credit_file(path : String) -> Dictionary:
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:43]: Please see other comment on method naming
```suggestion
func load_credit_file(path: String) -> Dictionary:
```

File: game/src/Credits/Credits.gd [hunk 3]
```diff
@@ -32,72 +32,85 @@ var label_settings_personel : LabelSettings
...
+		if roles[title_str].size() > 1:
+			push_warning("More than one %s: %s in %s" % [title_str, roles[title_str], path])
+			roles[title_str] = [roles[title_str][0]]
+	else:
+		push_warning("Credits file %s missing %s" % [path, title_str])
+	for role_list in roles.values():
+		role_list.sort_custom(func(a : String, b : String) -> bool: return a.naturalnocasecmp_to(b) < 0)
 	return roles
 
+func _add_label(node : Node, text : String, settings : LabelSettings) -> void:
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:85]:
```suggestion
func add_label(node: Node, text: String, settings: LabelSettings) -> void:
```

File: game/src/Credits/Credits.gd [hunk 4]
```diff
@@ -32,72 +32,85 @@ var label_settings_personel : LabelSettings
...
 
 		for person in project[role]:
-			var person_label = Label.new()
-
-			person_label.name = 'Label' + person
-			person_label.text = person
-			person_label.horizontal_alignment = HORIZONTAL_ALIGNMENT_CENTER
-			person_label.label_settings = label_settings_personel
-			role_parent.add_child(person_label)
+			_add_label(role_parent, person, label_settings_personel)
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:111]:
```suggestion
            add_label(role_parent, person, label_settings_personel)
```

File: game/src/Credits/Credits.gd [hunk 5]
```diff
@@ -32,72 +32,85 @@ var label_settings_personel : LabelSettings
...
 
-		project_credits_list.add_child(role_label)
+		_add_label(project_credits_list, role, label_settings_role)
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:113]:
```suggestion
        add_label(project_credits_list, role, label_settings_role)
```

File: game/src/Credits/Credits.gd [hunk 6]
```diff
@@ -106,11 +119,10 @@ func make_project_credits(project):
 # REQUIREMENTS:
 # * SS-17
 func _ready():
-	var core_credits = load_credit_file(core_credits_path)
-	make_project_credits(core_credits)
+	_add_project_credits(_load_credit_file(core_credits_path))
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:122]:
```suggestion
    add_project_credits(load_credit_file(core_credits_path))
```

File: game/src/Credits/Credits.gd [hunk 7]
```diff
@@ -32,72 +32,85 @@ var label_settings_personel : LabelSettings
...
-	project_credits_list.name = 'Credits' + project['TITLE'][0]
-	project_credits_label.name = 'Label' + project['TITLE'][0]
-	project_credits_label.text = project['TITLE'][0]
-	project_credits_label.horizontal_alignment = HORIZONTAL_ALIGNMENT_CENTER
-	project_credits_label.label_settings = label_settings_project
+	project_credits_list.name = 'Credits'
+	if title_str in project:
+		var title : String = project[title_str][0]
+		project_credits_list.name += title
+		_add_label(project_credits_list, title, label_settings_project)
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:101]:
```suggestion
        add_label(project_credits_list, title, label_settings_project)
```

File: game/src/Credits/Credits.gd [hunk 8]
```diff
@@ -32,72 +32,85 @@ var label_settings_personel : LabelSettings
 @export
 var credits_list: VBoxContainer
 
+const title_str : String = "TITLE"
```
Comment (Orwellian-225) [inline game/src/Credits/Credits.gd:35]: Adding str doesn't really explain what its for

```suggestion
const title_key: String = "TITLE"
```

**Review by Orwellian-225 — APPROVED** [review]

File: game/src/Credits/Credits.gd (hunk 1 above)
Comment (Hop311) [reply game/src/Credits/Credits.gd:95]: > Prepend a single underscore (_) to virtual methods functions the user must override, private functions, and private variables

I was using the underscore in the private function sense here, mainly so I could assume the Dictionary argument is formatted correctly (no empty list or non-list values).

Comment (Orwellian-225) [reply game/src/Credits/Credits.gd:95]: ah ok, makes sense. Then ignore the rest of my comments on here

File: game/src/Credits/Credits.gd (hunk 8 above)
Comment (Hop311) [reply game/src/Credits/Credits.gd:35]: I agree `title_key` is a better name. On whether to have a space before the colon in typed arguments, the only file in the repository currently using the "no space" style is `Options.gd`, so I'll leave that to a wider formatting PR.

File: game/src/Credits/Credits.gd (hunk 1 above)
Comment (Hop311) [reply game/src/Credits/Credits.gd:95]: No worries, in practice private functions are often connected to signals. Although the underscore doesn't actually prevent functions and variables from being accessed externally, it does stop them from appearing in the list of autocomplete options in the Godot editor, and generally makes it clear that you shouldn't be using/accessing the function/variable unless you know what you're doing.

File: game/src/Credits/Credits.gd (hunk 8 above)
Comment (Orwellian-225) [reply game/src/Credits/Credits.gd:35]: Yeah sure, no space thing doesn't matter that much to me just thought I'd mention it

**Review by ClarkeCode — APPROVED** [review]

**Pull request merged**